src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,439 @@
+        return true;
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return true;
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/* Input size directly since public key type structures may be aligned to word boundary */
+static size_t ec_pub_key_size(const char* curve, bool compressed)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(256) : EC_PUB_UNCOMPRESSED_SIZE(256));
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(384) : EC_PUB_UNCOMPRESSED_SIZE(384));
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return sizeof(soter_container_hdr_t)
+               + (compressed ? EC_PUB_SIZE(521) : EC_PUB_UNCOMPRESSED_SIZE(521));
+    } else {
+        return 0;
+    }
+}
+
+static size_t ec_priv_key_size(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return sizeof(soter_ec_priv_key_256_t);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return sizeof(soter_ec_priv_key_384_t);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return sizeof(soter_ec_priv_key_521_t);
+    } else {
+        return 0;
+    }
+}
+
+static char* ec_pub_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PUB_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PUB_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PUB_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static char* ec_priv_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PRIV_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PRIV_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PRIV_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static size_t bn_encode(const BIGNUM* bn, uint8_t* buffer, size_t length)
+{
+    int bn_size = BN_num_bytes(bn);
+    if (length < (size_t)bn_size) {
+        return 0;
+    }
+    memset(buffer, 0, length - bn_size);
+    return (length - bn_size) + BN_bn2bin(bn, buffer + (length - bn_size));
+}
+
+soter_status_t soter_engine_specific_to_ec_pub_key(const soter_engine_specific_ec_key_t* engine_key,
+                                                   bool compressed,
+                                                   soter_container_hdr_t* key,
+                                                   size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
```

Comment:
Would `const EVP_PKEY*` work, btw? Since the original pointer is to a const object.

Reply:
Well, yes, but in some places (such as `EVP_PKEY_set_utf8_string_param()`) usual `EVP_PKEY*` is expected, even though it doesn't seem to be modified. _Or is it? Theoretically, some caching may happen inside._
Not the first time non-const argument is expected when it's not actually modified.
To avoid warning and unnecessary casting back to `EVP_PKEY*`, decided to leave like this for now.

Reply:
I'd leave a comment here, at least with "FIXME: deal with this mess", since if you're intending to modify a key then this function should not accept a `const` pointer to begin with. But that's too big of a refactoring that won't really make things better, only "technically correct".

But what I'd really recommend is to look for a way to avoid that EVP_PKEY_set_utf8_string_param() call, since it's a bit crazy that exporting key material requires the key to be mutable. Maybe you're not using the right API here?

Reply:
Since I have to switch point format before extraction (in `3.0.8` it defaults to `uncompressed`), there seems to be no way to avoid mutating `pkey`. Well, one possible way is to set its `point-format` to `compressed` somewhere before (right after key generation?). But that seems dirty as some other unrelated place will decide whether the point will be compressed or not in this function.

> Looks like it will be possible to leave it const if OpenSSL 3.0.8 is the minimum supported version.

Sure? If `pub` contained compressed point by default in 3.0.7 and in 3.0.8 switched to uncompressed so newer version requires setting a different format.